# peterk87/nf-ionampliseq

Read mapping, variant calling and consensus sequence generation workflow for Ion Torrent Ampliseq sequence data of [FMDV] and [CSFV].

> **NB:** Built-in Ion Torrent AmpliSeq panels for Zika virus ([ZIKV]), Ebola virus ([EBOV]) and SARS-CoV-2 will be added shortly!

[![GitHub Actions CI Status](https://github.com/peterk87/nf-ionampliseq/workflows/CI/badge.svg)](https://github.com/peterk87/nf-ionampliseq/actions)
[![GitHub Actions Linting Status](https://github.com/peterk87/nf-ionampliseq/workflows/nf-core%20linting/badge.svg)](https://github.com/peterk87/nf-ionampliseq/actions)
[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A519.10.0-brightgreen.svg)](https://www.nextflow.io/)

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](https://bioconda.github.io/)
[![Docker](https://img.shields.io/docker/automated/peterk87/nf-ionampliseq.svg)](https://hub.docker.com/r/peterk87/nf-ionampliseq)

## Introduction

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It comes with docker containers making installation trivial and results highly reproducible.

## Quick Start

1. Install [`nextflow`](https://nf-co.re/usage/installation)

2. Install either [`Docker`](https://docs.docker.com/engine/installation/) or [`Singularity`](https://www.sylabs.io/guides/3.0/user-guide/) for full pipeline reproducibility _(please only use [`Conda`](https://conda.io/miniconda.html) as a last resort; see [docs](https://nf-co.re/usage/configuration#basic-configuration-profiles))_

3. Download the pipeline and test it on a minimal dataset with a single command:

    ```bash
    nextflow run peterk87/nf-ionampliseq -profile test,<docker/singularity/conda/institute>
    ```

    > Please check [nf-core/configs](https://github.com/nf-core/configs#documentation) to see if a custom config file to run nf-core pipelines already exists for your Institute. If so, you can simply use `-profile <institute>` in your command. This will enable either `docker` or `singularity` and set the appropriate execution settings for your local compute environment.

4. Start running your own analysis!

    <!-- TODO nf-core: Update the example "typical command" below used to run the pipeline -->

    ```bash
    nextflow run peterk87/nf-ionampliseq -profile <docker/singularity/conda/institute> --input '/path/to/iontorrent/*.bam'
    ```

See [usage docs](docs/usage.md) for all of the available options when running the pipeline.

## Documentation

The peterk87/nf-ionampliseq pipeline comes with documentation about the pipeline which you can read at [https://peterk87/nf-ionampliseq/docs](https://peterk87/nf-ionampliseq/docs) or find in the [`docs/` directory](docs).

This workflow includes several built-in analysis packages for Ion Torrent AmpliSeq sequence data of [CSFV] and [FMDV]. Users can also specify their own analysis packages, however, these files must be compatible with the Ion Torrent Software Suite including [tmap] and [tvc]

### Input

There are three methods of specifying input: `--input`; `--rundir`; `--sample_sheet` and `--panel` (or `--ref_fasta` and `--bed_file`). For all input modes, the primary input is BAM files generated by [Torrent Suite][] [tmap][] tooling.

The simplest way of running this workflow is with `--input` pointing at your Ion Torrent [Torrent Suite] produced BAM files:

```bash
nextflow run peterk87/nf-ionampliseq -profile <docker/singularity> --input '/path/to/*.bam'
```

With BAM file inputs specified via `--input`, the sample name and correct AmpliSeq panel (either [CSFV] or [FMDV]) will be determined from the BAM file headers.

You can also specify the Ion Torrent sequencing run directory as input with `--rundir`. All BAM files matching `IonCode_*_rawlib.bam` will be run through the workflow with sample names retrieved from the `ion_params_00.json`.

```bash
nextflow run peterk87/nf-ionampliseq -profile <docker/singularity> --rundir /path/to/rundir
```

If you wish to use custom names and a specific AmpliSeq panel it is recommended that you specify the following:

- `--sample_sheet`
  - CSV file with 2 columns:
    - Column 1: sample name
    - Column 2: path to raw BAM file from Ion Torrent (absolute path recommended)
- `--panel`
  - either `csf` or `fmd` for built-in AmpliSeq panel, otherwise, the user will need to specify a reference genome(s) FASTA file (`--ref_fasta`) and detailed BED file (`--bed_file`)

### Steps

1. [BAM Sample Info](docs/output.md#bam-sample-info) - Sample info extracted from BAM file headers
2. [FASTQ Reads](#fastq-reads) - BAM to FASTQ output
3. [FastQC](docs/output.md#fastqc) - Read quality control
4. [Mash](docs/output.md#mash) - Top reference genome determination by [Mash][] screen
5. [TMAP](docs/output.md#tmap) - Read mapping using the Thermo Fisher mapper [tmap]
6. [Samtools](docs/output.md#samtools) - Read mapping stats calculation with [Samtools][]
7. [Mosdepth](docs/output.md#mosdepth) - Coverage stats calculated by [Mosdepth][]
8. [TVC](docs/output.md#tvc) - Variant calling using the Thermo Fisher variant caller [tvc]
9. [Bcftools](docs/output.md#bcftools) - Variant filtering for majority consensus sequence generation and variant statistics for MultiQC report.
10. [Consensus Sequence](docs/output.md#consensus-sequence) - Majority consensus sequence with `N` masking of low/no coverage positions.
11. [Edlib Pairwise Alignment](docs/output.md#edlib-pairwise-alignment) - Pairwise global alignment and edit distance between reference and consensus sequences.
12. [Coverage Plots](#coverage-plots) - Coverage plots with/without low/no coverage and/or variants highlighted with linear and log10 scaling of y-axis depth values.
13. [MultiQC](docs/output.md#multiqc) - Aggregate report describing results from the whole pipeline. Consensus sequences are embedded in the MultiQC HTML report and can be downloaded from it.
14. [Pipeline information](docs/output.md#pipeline-information) - Report metrics generated during the workflow execution

### Output

For more information about the analysis steps and output of the pipeline, see the [output documenation](docs/output.md).

## Credits

peterk87/nf-ionampliseq was originally written by Peter Kruczkiewicz.

## Contributions and Support

If you encounter any issues when running this pipeline, please see the documentation above

If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).

The development of this pipeline tries to follow the guidelines and best-practices established by [nf-core](https://nf-co.re/) and was bootstrapped using [nf-core tools](https://nf-co.re/tools#creating-a-new-workflow). One day this pipeline may be added to nf-core.

## Citation

<!-- TODO nf-core: Add citation for pipeline after first release. Uncomment lines below and update Zenodo doi. -->
<!-- If you use  peterk87/nf-ionampliseq for your analysis, please cite it using the following doi: [10.5281/zenodo.XXXXXX](https://doi.org/10.5281/zenodo.XXXXXX) -->

You can cite the `nf-core` publication as follows:

> **The nf-core framework for community-curated bioinformatics pipelines.**
>
> Philip Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso & Sven Nahnsen.
>
> _Nat Biotechnol._ 2020 Feb 13. doi: [10.1038/s41587-020-0439-x](https://dx.doi.org/10.1038/s41587-020-0439-x).
> ReadCube: [Full Access Link](https://rdcu.be/b1GjZ)

<!-- External links and references -->

[bcftools]: https://samtools.github.io/bcftools/bcftools.html
[Bcftools]: https://samtools.github.io/bcftools/bcftools.html
[CSFV]: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi
[EBOV]: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=186536&lvl=3&lin=f&keep=1&srchmode=1&unlock
[Edlib]: https://github.com/Martinsos/edlib
[FMDV]: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=12110&lvl=3&lin=f&keep=1&srchmode=1&unlock
[Mash]: https://doi.org/10.1186/s13059-019-1841-x
[Mosdepth]: https://github.com/brentp/mosdepth
[Samtools]: https://www.htslib.org/
[tmap]: https://github.com/iontorrent/TS/
[TMAP]: https://github.com/iontorrent/TS/
[Torrent Suite]: https://github.com/iontorrent/TS
[tvc]: http://updates.iontorrent.com/tvc_standalone/
[TVC]: http://updates.iontorrent.com/tvc_standalone/
[variantCaller]: https://github.com/iontorrent/TS/tree/master/plugin/variantCaller
[vcf_consensus_builder]: https://github.com/peterk87/vcf_consensus_builder
[ZIKV]: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=64320&lvl=3&lin=f&keep=1&srchmode=1&unlock
